### ISSUE TITLE
Pagination seems a bit ... confusing (specifically Datastore)

### ISSUE THREAD

**Issue by jgeewax** [GoogleCloudPlatform/gcloud-ruby#707]
Doc: https://googlecloudplatform.github.io/gcloud-ruby/#/docs/v0.10.0/gcloud/datastore#paginating-records

This discussion has also been happening over at https://github.com/GoogleCloudPlatform/gcloud-python/issues/895

Basic gist is that we have a few different scenarios when "listing things":

1. Give me all the things
  - Kind of link infiniscroll
  - I don't care about how many API calls are made under the hood
  - I don't care about how many items there are

1. Give me all the things, up to N
  - Same as above, but stop at N things

1. Give me all the things, but no more than N API calls
  - Same as (1) except limit this to 100 API calls so I don't pay out the ass...

1. Give me page X of things, where each page is exactly N items (unless I'm out of items)
  - I start with no token, which gives me "page 1"
  - I get a token with that response, which I can pass in later to get "the next page"
  - The only reason why I should get < N items in a page is if there are no more to give me

---

Some examples of what I'd want... (assumes we start with a Datastore `query` object):

```ruby
# Give me all the things forever
results = dataset.run query
results.each do |entity|
  puts "Got entity #{entity}"
end

# And in another list method....
topics = pubsub.topics
topics.each do |topic|
  puts "Got topic: #{topic.name}"
end

# Give me up to 100 things
first_100_topics = pubsub.topics limit: 100
topics.each do |topic|
  puts "Got topic: #{topic.name}"
end

# Give me all the things up to 100 API calls
topics = pubsub.topics max_api_calls: 100  # Maybe a different name?

# Give me paginated things.
first_page_of_topics = pubsub.topics page_size: 10
first_page_of_topics.each do |topic|
  puts "Got topic: #{topic.name}"
end

# Give me the next page of 10
next_page_of_topics = pubsub.topics page_size: 10, page_token: first_page_of_topics.next_page_token
first_page_of_topics.each do |topic|
  puts "Got topic: #{topic.name}"
end
```

**Comment by quartzmo** [comment]
@jgeewax I edited the docs link in your description to point to `googlecloudplatform.github.io` (was `quartzmo` fork.) Thanks for all these suggestions! I'll keep an eye on GoogleCloudPlatform/gcloud-python#895.

**Comment by blowmage** [comment]
I think we can get most of the way there with lazy enumerables. See #690. The only thing that won't do is limit the number of total API calls.

**Comment by quartzmo** [comment]
@jgeewax PR #708 (just merged) replaces the example you linked with this one:

```ruby
gcloud = Gcloud.new
datastore = gcloud.datastore

query = datastore.query("Task")
tasks = datastore.run query

loop do
  tasks.each do |t|
    puts t["description"]
  end
  break unless tasks.next?
  tasks = tasks.next
end
```

**Comment by blowmage** [comment]
We could also use `#all`, but it will pre-fetch all the results before starting the iteration. It isn't lazy, but it is in the current release:

```ruby
gcloud = Gcloud.new
datastore = gcloud.datastore

query = datastore.query("Task")
tasks = datastore.run query

tasks.all.each do |t|
  puts t["description"]
end
```

I'm currently working on an implementation of a lazy enumerator, so in the future the pre-fetch can be avoided and the results can be iterated while pages are fetched:

```ruby
gcloud = Gcloud.new
datastore = gcloud.datastore

query = datastore.query("Task")
tasks = datastore.run query

tasks.lazy.each do |t|
  puts t["description"]
end
```

Lazy enumerators are super cool, as you can also take the first N records, regardless of how many pages it takes to get them. And since it is lazy it won't ask for more results than are needed:

```ruby
gcloud = Gcloud.new
datastore = gcloud.datastore

query = datastore.query("Task")
tasks = datastore.run query

tasks.lazy.take(50).each do |t|
  puts t["description"]
end
```

I haven't implemented this yet, but I believe it is possible to also put a restriction on the number of max pages asked for:

```ruby
gcloud = Gcloud.new
datastore = gcloud.datastore

query = datastore.query("Task")
tasks = datastore.run query

tasks.lazy(max_api_calls: 10).take(50).each do |t|
  puts t["description"]
end
```

**Comment by jgeewax** [comment]
Sounds like a lazy enumerator is the way to go (http://patshaughnessy.net/2013/4/3/ruby-2-0-works-hard-so-you-can-be-lazy)

My big question then is... how do I tell you where to start the next page? Ie, imagine I'm building a web app where I want to see 10 items at a time, and click the "next page" link? How does that work?

**Comment by quartzmo** [comment]
@blowmage I think changing the return type of `#all` from an array to an enumerator that supports [`Enumerable#lazy`](http://ruby-doc.org/core-2.0.0/Enumerable.html#method-i-lazy) is the perfect solution, and appears compatible with the addition of an option for `max_api_calls `. Let's do this asap.

**Comment by blowmage** [comment]
Resolved in the 0.11.0 release.
